pkg/resource/deploy/step_generator.go

```diff
@@ -1300,6 +1314,57 @@ func applyReplaceOnChanges(diff plugin.DiffResult,
 	}, nil
 }
 
+func (sg *stepGenerator) addChild(parent resource.URN, child *resource.State) {
+	if parent != "" {
+		sg.children[parent] = append(sg.children[parent], child)
+	}
+}
+
+// addDependency adds a dependency on the given resource to the set of deps.
+//
+// The behavior of this method depends on whether or not the resource is a custom resource or a component resource:
+//
+// - Custom resources are added directly to the set.
+// - Component resources are added to the set, but also act as aggregations of their descendents, which are also
+//   added to the set.
+//
+// In other words, if we had a dependency on Comp1 in the following parent/child graph:
+//
+//               Comp1
+//          _______|__________
+//          |      |          |
+//       Cust1   Comp2      Comp3
+//              ___|___       |
+//              |     |       |
+//            Cust2  Cust3  Comp4
+//              |             |
+//            Cust4         Cust5
+//
+// Then the dependency set will be [Comp1, Cust1, Comp2, Comp3, Cust2, Cust3, Comp4, Cust5]. Notably, Cust4 is *not*
+// a member of the set because it is a child of a custom resource.
+func (sg *stepGenerator) addDependency(deps urnSet, res *resource.State) {
+	if !res.Custom {
+		for _, child := range sg.children[res.URN] {
+			sg.addDependency(deps, child)
+		}
```

Comment:
This assumes that the `sg.children` map cannot contain a cycle.  I'm not 100% sure what strictly prevents that from happening - specifically registering a resource with a parent that doesn't exist (yet), but which is registered in the future to close the cycle.  I suspect that is guaranteed to fail in some way, but I cannot see clearly where exactly that happens.  

Is it possible to have `addChild` enforce that there are no cycles so that this invariant is maintained more locally?

Reply:
> I suspect that is guaranteed to fail in some way, but I cannot see clearly where exactly that happens.

I believe that is guaranteed to fail as well, though I think that it's pretty far removed from the step generator. I _think_ that this actually happens in the snapshot manager when we're verifying snapshot integrity. Adding validation to either `addChild` or `GenerateSteps` seems reasonable to me. In the former case, we could do what you suggest; in the latter case, we could validate that a resource's parent actually exists.

Reply:
Added validation to `addChild` to ensure that a resource's parent exists.